xds/src/main/java/io/grpc/xds/EnvoyProtoData.java

```diff
@@ -135,17 +141,370 @@ public String toString() {
+        return this;
+      }
+
+      Node build() {
+        return new Node(
+            id, cluster, metadata, locality, listeningAddresses, buildVersion, userAgentName,
+            userAgentVersion, clientFeatures);
+      }
+    }
+
+    static Builder newBuilder() {
+      return new Builder();
+    }
+
+    Builder toBuilder() {
+      Builder builder = new Builder().setId(id).setCluster(cluster);
+      if (metadata != null) {
+        builder.setMetadata(metadata);
+      }
+      if (locality != null) {
+        builder.setLocality(locality);
+      }
+      builder.listeningAddresses.addAll(listeningAddresses);
+      return builder;
+    }
+
+    String getId() {
+      return id;
+    }
+
+    String getCluster() {
+      return cluster;
+    }
+
+    @Nullable
+    Map<String, ?> getMetadata() {
+      return metadata;
+    }
+
+    @Nullable
+    Locality getLocality() {
+      return locality;
+    }
+
+    List<Address> getListeningAddresses() {
+      return listeningAddresses;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+      Node node = (Node) o;
+      return Objects.equals(id, node.id)
+          && Objects.equals(cluster, node.cluster)
+          && Objects.equals(metadata, node.metadata)
+          && Objects.equals(locality, node.locality)
+          && Objects.equals(listeningAddresses, node.listeningAddresses)
+          && Objects.equals(buildVersion, node.buildVersion)
+          && Objects.equals(userAgentName, node.userAgentName)
+          && Objects.equals(userAgentVersion, node.userAgentVersion)
+          && Objects.equals(clientFeatures, node.clientFeatures);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(
+          id, cluster, metadata, locality, listeningAddresses, buildVersion, userAgentName,
+          userAgentVersion, clientFeatures);
+    }
+
+    static Node fromEnvoyProtoNode(io.envoyproxy.envoy.config.core.v3.Node proto) {
+      Node.Builder builder = newBuilder();
+      builder.setId(proto.getId());
+      builder.setCluster(proto.getCluster());
+      if (proto.hasMetadata()) {
+        builder.setMetadata(proto.getMetadata().getFieldsMap());
+      }
+      if (proto.hasLocality()) {
+        builder.setLocality(Locality.fromEnvoyProtoLocality(proto.getLocality()));
+      }
+      List<io.envoyproxy.envoy.config.core.v3.Address> listeningAddressesList =
+          proto.getListeningAddressesList();
+      List<Address> listeningAddresses = new ArrayList<>(listeningAddressesList.size());
+      for (io.envoyproxy.envoy.config.core.v3.Address addressProto : listeningAddressesList) {
+        listeningAddresses.add(Address.fromEnvoyProtoAddress(addressProto));
+      }
+      builder.listeningAddresses.addAll(listeningAddresses);
```

Comment:
Why this is not calling `builder.addListeningAddresses(...)` while looping through the list?

Reply:
Because we chose to have `Builder. addListeningAddresses(Address address)` instead of `Builder.setListeningAddresses(List<Address>)`, so builder.listeningAddresses has to be final, otherwise `Node.toBuilder().addListeningAddresses()` will be broken. 

Because `builder.listeningAddresses` is final, we need `addAll()`.

"`builder.addListeningAddresses(...)` while looping through the list" is also why I chose `setListeningAddresses` in my first version.

Reply:
I don't understand your reasoning...

What's the problem of "builder.addListeningAddresses(...) while looping through the list"? You are looping over proto messages to convert `Address`. What's the problem of calling `addListeningAddresses` in each iteration?

Reply:
I misunderstood you question. You were right. Fixed.